lib/src/layer/tile_layer/tile_provider/base_tile_provider.dart

```diff
@@ -1,84 +1,284 @@
+
+  /// Construct the base tile provider and initialise the [headers] property
+  ///
+  /// This is not a constant constructor, and does not use an initialising
+  /// formal, intentionally. To enable [TileLayer] to efficiently (without
+  /// [headers] being non-final or unstable `late`) inject the appropriate
+  /// 'User-Agent' (based on what is specified by the user), the [headers] [Map]
+  /// must not be constant.
+  ///
+  /// Implementers/extenders should call add `super.headers` to their
+  /// constructors if they support custom HTTP headers.
+  TileProvider({Map<String, String>? headers}) : headers = headers ?? {};
 
-  /// Retrieve a tile as an image, based on it's coordinates and the
-  /// current [TileLayerOptions]
-  ImageProvider getImage(TileCoordinates coordinates, TileLayer options);
+  /// Retrieve a tile as an image, based on its coordinates and the [TileLayer]
+  ///
+  /// Does not support cancelling loading tiles, unlike
+  /// [getImageWithCancelLoadingSupport]. For this method to be called instead of
+  /// that, the implementation of [TileProvider] must not override
+  /// [supportsCancelLoading] to `true`.
+  ///
+  /// Usually redirects to a custom [ImageProvider], with one input depending on
+  /// [getTileUrl].
+  ///
+  /// For many implementations, this is the only method that will need
+  /// implementing.
+  ImageProvider getImage(TileCoordinates coordinates, TileLayer options) {
+    throw UnimplementedError(
+      'A `TileProvider` that does not override `supportsCancelLoading` to `true` '
+      'must override `getImage`',
+    );
+  }
+
+  /// Retrieve a tile as an image, based on its coordinates and the [TileLayer]
+  ///
+  /// Supports cancelling loading tiles, which is designed to allow for
+  /// implementations that can cancel HTTP requests in-flight, when the
+  /// underlying tile is disposed before it is loaded. This may increase
+  /// performance, and may decrease unnecessary tile requests.
+  ///
+  /// The [cancelLoading] future will complete when the underlying tile is
+  /// disposed/pruned. The implementation should therefore listen for its
+  /// completion, then cancel the loading and return ([ImageDecoderCallback])
+  /// [transparentImage].
+  ///
+  /// For this method to be called instead of [getImage], the implementation of
+  /// [TileProvider] must override [supportsCancelLoading] to `true`.
+  ///
+  /// Usually redirects to a custom [ImageProvider], with one parameter using
+  /// [getTileUrl], and one using [cancelLoading].
+  ///
+  /// For many implementations, this is the only method that will need
+  /// implementing.
+  ImageProvider getImageWithCancelLoadingSupport(
+    TileCoordinates coordinates,
+    TileLayer options,
+    Future<void> cancelLoading,
+  ) {
+    throw UnimplementedError(
+      'A `TileProvider` that overrides `supportsCancelLoading` to `true` must '
+      'override `getImageWithCancelLoadingSupport`',
+    );
+  }
```

Comment:
Wouldn't it be possible to merge `.getImage()` and `getImageWithCancelLoadingSupport()` into one function? The implementation will probably only override one function anyways and could handle additional logic inside the function.
Your solution is probably best for backwards compatibility but it feels a bit odd to me.

Reply:
Yeah, I was aiming for backwards compatibility. I have made breaking changes in this PR, but most implementations won't be affected.
If the other maintainers think that this isn't necessary, I'm happy to simplify it.